run-as.1
.\" (c) 2014 Richard Kettlewell
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.TH run-as 1
.SH NAME
run-as \- run a command as a specified user
.SH SYNOPSIS
.B run-as
.RI [ options ]
.RB --
.I user
.RI [ group ]
.B --
.IR command ...
.SH DESCRIPTION
\fBrun-as\fR runs a command as a specified user.
.PP
If only \fIuser\fR is specified then the default group for that user
is assumed.  If \fIgroup\fR is specified as well then that group is
used.
.SH OPTIONS
.TP
\fB-r\fR \fIdirectory\fR, \fB--root\fR \fIdirectory\fR
Alter the root directory before changing user.
.TP
\fB-h\fR, \fB--help\fR
Show summary of options.
.TP
\fB-V\fR, \fB--version\fR
Show version of program.
.SH NOTES
The \fB--\fR arguments separating the options from the user and the
user (and group) from the command are mandatory.
.PP
Although it would seem more logical to have separate programs for
changing user and changing root directory, this can't be done.  If it
were done in separate programs then the change root program wouldn't
work if the change user program had already been run (since only
privileged users can change the root directory of a process).  On the
other hand if the change root program were run first, the change user
program would no longer have access to the files necessary to look up
the user information.
.PP
If \fB--root\fR is used, then the command must be available inside the
chroot.  If this is a problem then you must write your program to
start as a privileged user and change its root directory and UID
itself.
.SH "SEE ALSO"
.BR chroot (2)
.SH AUTHOR
Richard Kettlewell <rjk@greenend.org.uk>


ChangeLog
2005-02-07 17:17:21 +0000  Richard Kettlewell  <rjk@greenend.org.uk>

        * adverbio.c: -O FD:PATH where the open(PATH, ...) returned a value
          matching FD would unhelpfuly close FD.  Problem reported by Colin
          Watson.
2005-02-07 17:14:08 +0000  Richard Kettlewell  <rjk@greenend.org.uk>

        * with-lock.c: bogus string literal
2003-07-19 17:35:12  Richard Kettlewell  <rjk@greenend.org.uk>

        * check-tested-everything, test-*: check that every executable has
          been tested
        
        * configure.ac: quieten autoconf
        
        * debug.c, utils.h, run-repeatedly.c: new debugging interface
        
        * test-run-repeatedly: fix test
        
2003-07-19 11:26:29  Richard Kettlewell  <rjk@greenend.org.uk>

        * prepare: don't configure if asked not to
        
        * test-run-repeatedly: tests
        
2003-03-08  Richard Kettlewell  <rjk@greenend.org.uk>

	* inplace.1: added some examples
	
2002-11-28  Richard Kettlewell  <rjk@greenend.org.uk>

	* iobuffer.c: use a single big circular buffer rather than lots of
	small buffers.  This makes more predictable and efficient use of
	memory than the previous strategy.

	Also change --max-buffer/-m to --buffer/-b since it is now the
	exact buffer size.

2002-11-23  Richard Kettlewell  <rjk@greenend.org.uk>

	* More test infrastructure.  Various tests improved.

	* run-repeatedly.c: underscore in filenames is allowed (to match
	run-parts)

2002-11-22  Richard Kettlewell  <rjk@greenend.org.uk>

	* various changes to build on FreeBSD.  daemon (at least) doesn't
	seem to work properly there though.

2002-11-21  Richard Kettlewell  <rjk@greenend.org.uk>

	* alarm.c: report failed setpgid correctly

	* lots of <syscall>_e functions to keep error handling out of the
	main code.  Convert a couple of malloc/realloc calls to the x-
	versions too (for the same reason).

	* more tests

2002-11-20  Richard Kettlewell  <rjk@greenend.org.uk>

	* test-daemon: more tests, including for some of the new features

	* logfds.c: adjust option letters so daemon & logfds are
	consistent.

	* daemon.c: saner error output handling - essentially build part
	of logfds into daemon.

2002-11-18  Richard Kettlewell  <rjk@greenend.org.uk>

	* anagrams.c: use dynamic memory allocation rather than fixed-size
	buffers.  Now there should be no size limits other than available
	memory.

	Also, now you can specify multiple word lists, and duplicate words
	don't cause duplicate anagrams.

2002-11-17  Richard Kettlewell  <rjk@greenend.org.uk>

	* new programs: iobuffer, anagrams

2002-11-11  Richard Kettlewell  <rjk@greenend.org.uk>

	* inplace.c: fix message reported when a subprocess gets a signal

	* alarm: new --verbose option

	* new programs: logfds, bind/accept/connect-socket, pidfile,
	run-as, daemon, run-repeatedly, with-lock

	* switch to new autoconf &c

2001-09-09  Richard Kettlewell  <rjk@greenend.org.uk>

	* configure.in: changed version to 0.6.0

	* added "daemon" program.

2001-08-12  Richard Kettlewell  <rjk@greenend.org.uk>

	* merged "inplace" and "adverbio" into a single package



run-as.c
/* 
   run-as - run a command as a specified user

   Copyright (C) 2001, 2002 Richard Kettlewell

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include <config.h>

#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <getopt.h>

#include "utils.h"

/* Option flags and variables */
static struct option const long_options[] =
{
  { "help", no_argument, 0, 'h' },
  { "version", no_argument, 0, 'V' },
  { "root", required_argument, 0, 'r' },
  { 0, 0, 0, 0}
};

/* write a usage message to FP and exit with the specified status */

static void __attribute__((noreturn)) usage(FILE *fp, int exit_status) {
  if(fputs(
"Usage:\n"
"  run-as [options] -- user [group] -- command ...\n"
"\n"
"Options:\n"
"  -r DIRECTORY, --root DIRECTORY    Change root directory\n"
"  -h, --help                        Usage message\n"
"  -V, --version                     Version number\n"
, fp) < 0)
    fatale("output error");
  exit(exit_status);
}

int main(int argc, char **argv) {
  int n;
  const char *root = 0;
  const char *user, *group;

  setprogname(argv[0]);
  
  while((n = getopt_long(argc, argv, 
			 "hVr:",
			 long_options, (int *)0)) >= 0) {
    switch(n) {
    case 'V':
      printf("run-as %s\n", VERSION);
      return 0;

    case 'h':
      usage(stdout, 0);

    case 'r':
      root = optarg;
      break;
      
    default:
      usage(stderr, 1);
    }
  }
  
  /* check for a user */
  if(optind >= argc)
    fatal("no user specified");
  user = argv[optind++];

  /* check for a group */
  if(optind < argc && strcmp(argv[optind], "--"))
    group = argv[optind++];
  else
    group = 0;

  /* check for the separator */
  if(optind >= argc || strcmp(argv[optind], "--"))
    fatal("missing command separator");
  ++optind;

  /* check there is a command */
  if(optind >= argc)
    fatal("missing command");

  /* do all the changes */
  setpriv(user, group, root);
  
  /* execute the command */
  execvp(argv[optind], argv + optind);
  fatale("error executing \"%s\"", argv[optind]);
}

/*
Local Variables:
c-basic-offset:2
comment-column:40
End:
*/
